risc0/zkvm/src/groth16.rs

```diff
@@ -0,0 +1,798 @@
+        for sub_vec in &self.b {
+            for item in sub_vec {
+                result.extend(item);
+            }
+        }
+
+        // Serialize 'c'
+        for item in &self.c {
+            result.extend(item);
+        }
+
+        result
+    }
+
+    /// Method to convert back from Vec<u8>
+    pub fn from_vec(data: &[u8]) -> Result<Groth16Seal, Error> {
+        if data.len() != Self::SIZE {
+            bail!("Data length mismatch");
+        }
+
+        let mut offset = 0;
+        let mut a = Vec::with_capacity(2);
+        let mut b = Vec::with_capacity(2);
+        let mut c = Vec::with_capacity(2);
+
+        // Deserialize 'a'
+        for _ in 0..2 {
+            a.push(data[offset..offset + Self::ELEMENT_SIZE].to_vec());
+            offset += Self::ELEMENT_SIZE;
+        }
+
+        // Deserialize 'b'
+        for _ in 0..2 {
+            let mut sub_vec = Vec::with_capacity(2);
+            for _ in 0..2 {
+                sub_vec.push(data[offset..offset + Self::ELEMENT_SIZE].to_vec());
+                offset += Self::ELEMENT_SIZE;
+            }
+            b.push(sub_vec);
+        }
+
+        // Deserialize 'c'
+        for _ in 0..2 {
+            c.push(data[offset..offset + Self::ELEMENT_SIZE].to_vec());
+            offset += Self::ELEMENT_SIZE;
+        }
+
+        Ok(Groth16Seal { a, b, c })
+    }
+}
+
+#[cfg(not(target_os = "zkvm"))]
+impl TryFrom<CircomProof> for Groth16Seal {
+    type Error = Error;
+    fn try_from(circom_proof: CircomProof) -> Result<Self, Error> {
+        if circom_proof.pi_a.len() < 2 {
+            bail!("Malformed G1 element field");
+        }
+        let a = vec![
+            from_u256(&circom_proof.pi_a[0])?,
+            from_u256(&circom_proof.pi_a[1])?,
+        ];
+
+        if circom_proof.pi_b.len() < 2
+            || circom_proof.pi_b[0].len() < 2
+            || circom_proof.pi_b[1].len() < 2
+        {
+            bail!("Malformed G2 element field");
+        }
+        let b = vec![
+            vec![
+                from_u256(&circom_proof.pi_b[0][1])?,
+                from_u256(&circom_proof.pi_b[0][0])?,
+            ],
+            vec![
+                from_u256(&circom_proof.pi_b[1][1])?,
+                from_u256(&circom_proof.pi_b[1][0])?,
+            ],
+        ];
+
+        if circom_proof.pi_c.len() < 2 {
+            bail!("Malformed G1 element field");
+        }
+        let c = vec![
+            from_u256(&circom_proof.pi_c[0])?,
+            from_u256(&circom_proof.pi_c[1])?,
+        ];
+
+        Ok(Groth16Seal { a, b, c })
+    }
+}
+
+/// Groth16 instance over the BN_254 curve encoded in little endian
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Groth16Proof {
+    pvk: Vec<u8>,
+    proof: Vec<u8>,
+    prepared_inputs: Vec<u8>,
+}
+
+impl Groth16Proof {
+    /// Creates a Groth16 instance from a `Groth16Seal` and the metadata digest
+    /// of the original RISC Zero receipt
+    #[cfg(not(target_os = "zkvm"))]
```

Comment:
We should avoid `#cfg[]` all over the place and instead mark this entire file as being 'host-only'.